closed/src/java.base/share/classes/openj9/internal/security/RestrictedSecurity.java

```diff
@@ -1043,10 +1049,20 @@ boolean isRestrictedProviderAllowed(String providerName) {
                 providerName = providerName.substring(0, pos);
             }
 
-            // Provider name defined in provider construction method.
-            providerName = getProvidersSimpleName(providerName);
+            // For the provider which provider name is different with its class
+            // name. First, check if the provider class name is in the restricted
+            // security provider list without removing the package name.
+            if (providersSimpleName.contains(providerName)) {
+                if (debug != null) {
+                    debug.println("The provider " + providerName + " class is allowed in restricted security mode.");
+                }
+                return true;
+            }
+
+            // Provider name without package name.
+            providerName = getProvidersNameWithoutPackage(providerName);
```

Comment:
I don't think we want this any more. Only the name as specified in the properties file (less any argument) should be used. The name of the field, `providersSimpleName`, should be changed: perhaps `providers` and `providersSimpleName`
should be `providerWithArg` and `providerWithoutArg`, respectively.

Reply:
@keithc-ca 

I remember in our last meeting, for the NSS FIPS 140-2 provider list, we still use the provider name (the name from its constructor) in the provider list. Only the OpenJCEPlus FIPS 140-3 and any new providers which need to be added into the NSS FIPS 140-2 provider list, for example the SunSASL, will use the provider fully qualified name.

```
RestrictedSecurity.NSS.140-2.jce.provider.1 = SunPKCS11 ${java.home}/conf/security/nss.fips.cfg
RestrictedSecurity.NSS.140-2.jce.provider.2 = SUN [{CertificateFactory, X.509, ImplementedIn=Software}, \
    {CertStore, Collection, ImplementedIn=Software}, \
    {CertStore, com.sun.security.IndexedCollection, ImplementedIn=Software}, \
    {Policy, JavaPolicy, *}, {Configuration, JavaLoginConfig, *}, \
    {CertPathBuilder, PKIX, ValidationAlgorithm=RFC5280:ImplementedIn=Software}, \
    {CertPathValidator, PKIX, ValidationAlgorithm=RFC5280:ImplementedIn=Software}, \
    {KeyStore, PKCS12, *}]
RestrictedSecurity.NSS.140-2.jce.provider.3 = SunEC [{KeyFactory, EC, ImplementedIn=Software: \
    SupportedKeyClasses=java.security.interfaces.ECPublicKey|java.security.interfaces.ECPrivateKey: \
    KeySize=256}, {AlgorithmParameters, EC, *}]
RestrictedSecurity.NSS.140-2.jce.provider.4 = SunJSSE
```

So, for example, the SunEC, when the ServiceLoader check the sun.security.ec.SunEC, we still need to get its provider name SunEC from removing the package. So, that is why I did not remove this getProvidersNameWithoutPackage() method.

@jasonkatonica FYI.

Reply:
My understanding is that we agreed that only fully-qualified class names would be allowed when specifying a restricted security profile (otherwise the restrictions can be bypassed). Yes, this may mean some migration work for users, but the longer we wait to make that change, the greater number of users will be impacted.